README.md

# ObliqueBART

<!-- badges: start -->
<!-- badges: end -->

The goal of ObliqueBART is to provide an implementation of Bayesian Additrive Regression Trees with Oblique splits (i.e. hyperplane splits defined by linear combinations of voariates)

## Installation

You can install the development version of ObliqueBART like so:

``` r
library(devtools)
install_github("EoghanONeill/ObliqueBART")
```

## Example

This is a basic example:

``` r
library(ObliqueBART)
## basic example code


# example without hyperprior
library(dbarts)
library(dplyr)

# Simulate data -------------------------------------------
f <- function(x) {
  10 * sin(pi * x[,1] * x[,2]) + 20 * (x[,3] - 0.5)^2 +
    10 * x[,4] + 5 * x[,5]
}

set.seed(99)
sigma <- 1.0
n     <- 200

x  <- matrix(runif(n * 10), n, 10)
Ey <- f(x)
y  <- rnorm(n, Ey, sigma)

# run BART --------------------------------------------------------
set.seed(99)
bartFit <- ObliqueBART(x, y, ntree = 10, nburn = 1000, npost = 5000)



cbind(apply(bartFit$y_hat,2,mean) , y)



ntest     <- 200

xtest  <- matrix(runif(ntest * 10), ntest, 10)
Eytest <- f(xtest)
ytest  <- rnorm(ntest, Eytest, sigma)


testres <- predict_mybart(bartFit, xtest,
               type = 'all')

cbind(apply(testres,2,mean) , ytest)



#### example of how the hyperprior can learn the important covariates


bartFit <- ObliqueBART(x, y, ntree = 10, nburn = 10000, npost = 10000,
                       coef_prior = "univariate_normal",
                       coef_hyperprior = "univariate_normal_betabinomial_theta_j",
                       coef_norm_hyperprior = "fixed")

testres <- predict_mybart(bartFit, xtest,
               type = 'all')

cbind(apply(testres,2,mean) , ytest)

bartFit$hyp_par_list$theta_vec



```



NEWS.md
# ObliqueBART (development version)

* Initial CRAN submission.


R/aux_tree.R
# -------------------------------------------------------------------------#
# Description: this script contains auxiliar functions needed to update    #
# the trees with details and to map the predicted values to each obs       #
# -------------------------------------------------------------------------#

# 1. fill_tree_details: takes a tree matrix and returns the number of obs in each node in it and the indices of each observation in each terminal node
# 2. get_predictions: gets the predicted values from a current set of trees
# 3. get_children: it's a function that takes a node and, if the node is terminal, returns the node. If not, returns the children and calls the function again on the children
# 4. resample: an auxiliar function
# 5. get_ancestors: get the ancestors of all terminal nodes in a tree
# 6. update_s: full conditional of the vector of splitting probability.
# 7. get_number_distinct_cov: given a tree, it returns the number of distinct covariates used to create its structure

# Fill_tree_details -------------------------------------------------------

fill_tree_details = function(curr_tree, X) {

  # Collect right bits of tree
  tree_matrix = curr_tree$tree_matrix

  # Create a new tree matrix to overwrite
  new_tree_matrix = tree_matrix

  # Start with dummy node indices
  node_indices = rep(1, nrow(X))

  # For all but the top row, find the number of observations falling into each one
  for(i in 2:nrow(tree_matrix)) {

    # Get the parent
    curr_parent = as.numeric(tree_matrix[i,'parent'])

    # Find the split variable and value of the parent
    # split_var = as.numeric(tree_matrix[curr_parent,'split_variable'])
    split_coefs = as.numeric(tree_matrix[curr_parent,8:(ncol(tree_matrix))])
    split_val = as.numeric(tree_matrix[curr_parent, 'split_value'])

    # print("split_val= ")
    # print(split_val)

    if(is.na(split_val)){
      stop("split_val == NA")
    }
    if(any(is.na(split_coefs))){
      stop("split_coefs contains NA")
    }

    # Find whether it's a left or right terminal node
    left_or_right = ifelse(tree_matrix[curr_parent,'child_left'] == i,
                           'left', 'right')

    if(left_or_right == 'left') {
      # If left use less than condition
      # new_tree_matrix[i,'node_size'] = sum(X[node_indices == curr_parent,split_var] < split_val)
      # node_indices[node_indices == curr_parent][X[node_indices == curr_parent,split_var] < split_val] = i

      # print("ncol(X[node_indices == curr_parent,]")
      # print(ncol(X[node_indices == curr_parent,]))
      # print("split_coefs = " )
      # print(split_coefs)

      new_tree_matrix[i,'node_size'] <- sum( ( X[node_indices == curr_parent,] %*% split_coefs )   < split_val )

      # print("new_tree_matrix[i,'node_size'] = ")
      # print(new_tree_matrix[i,'node_size'])

      node_indices[node_indices == curr_parent][ (X[node_indices == curr_parent,] %*% split_coefs ) < split_val ] <-  i



    } else {
      # If right use greater than condition
      # new_tree_matrix[i,'node_size'] = sum(X[node_indices == curr_parent,split_var] >= split_val)
      # node_indices[node_indices == curr_parent][X[node_indices == curr_parent,split_var] >= split_val] = i


      # # replaced >= with ! < because values equal or very close to equal to split vlaue not allocated to a node.
      # temp_vals <- X[node_indices == curr_parent,] %*% split_coefs
      # new_tree_matrix[i,'node_size'] <- sum(  !(( temp_vals ) < split_val)  |(temp_vals == split_val) )
      #
      # # print("new_tree_matrix[i,'node_size'] = ")
      # # print(new_tree_matrix[i,'node_size'])
      #
      # node_indices[node_indices == curr_parent][  !( ( temp_vals ) <  split_val ) |(temp_vals == split_val) ] <- i

      # right node always after left, so if not filled in as the left node, remaining indices are all the right node
      new_tree_matrix[i,'node_size'] <- sum(node_indices == curr_parent)
      node_indices[node_indices == curr_parent] <- i



    }
  } # End of loop through table

  # print("new_tree_matrix = ")
  # print(new_tree_matrix)
  # print("new_tree_matrix$terminal = ")
  # print(new_tree_matrix$terminal)

  # if(any( sort(unique(node_indices)) != which(new_tree_matrix[,"terminal"] == 1))){
  #
  #   indnoteq <- which( sort(unique(node_indices)) != which(new_tree_matrix[,"terminal"] == 1))
  #   missing_inds <- which(node_indices == indnoteq)
  #
  #
  #   print("node_indices = ")
  #   print(node_indices)
  #
  #   print("curr_parent = ")
  #   print(curr_parent)
  #
  #   print("new_tree_matrix = ")
  #   print(new_tree_matrix)
  #   print("  sort(unique(node_indices)) = ")
  #   print(  sort(unique(node_indices)) )
  #
  #   print("X[node_indices == curr_parent,] %*% split_coefs    = ")
  #   print(X[node_indices == curr_parent,] %*% split_coefs )
  #
  #   print("split_val = ")
  #   print(split_val)
  #
  #   print("!( X[node_indices == curr_parent,] %*% split_coefs   < split_val )  ")
  #   print(!(X[node_indices == curr_parent,] %*% split_coefs   < split_val ))
  #
  #   print("X[node_indices == curr_parent,] %*% split_coefs   < split_val  ")
  #   print(X[node_indices == curr_parent,] %*% split_coefs   < split_val )
  #
  #
  #   print("X[node_indices == curr_parent,] %*% split_coefs   - split_val  ")
  #   print(X[node_indices == curr_parent,] %*% split_coefs   - split_val )
  #
  #   # print("X[missing_inds,] %*% split_coefs )   = ")
  #   # print(X[missing_inds,] %*% split_coefs )
  #
  #
  #
  #   stop("any( sort(unique(node_indices)) != which(new_tree_matrix[,1] ==1))")
  # }


  return(list(tree_matrix = new_tree_matrix,
              node_indices = node_indices))

} # End of function

# Get predictions ---------------------------------------------------------

get_predictions = function(trees, X, single_tree = FALSE) {

  # Stop nesting problems in case of multiple trees
  if(is.null(names(trees)) & (length(trees) == 1)) trees = trees[[1]]

  # Normally trees will be a list of lists but just in case
  if(single_tree) {
    # Deal with just a single tree
    if(nrow(trees$tree_matrix) == 1) {
      predictions = rep(trees$tree_matrix[1, 'mu'], nrow(X))
    } else {
      # Loop through the node indices to get predictions
      predictions = rep(NA, nrow(X))
      unique_node_indices = unique(trees$node_indices)
      # Get the node indices for the current X matrix
      curr_X_node_indices = fill_tree_details(trees, X)$node_indices
      # Now loop through all node indices to fill in details
      for(i in 1:length(unique_node_indices)) {
        predictions[curr_X_node_indices == unique_node_indices[i]] =
          trees$tree_matrix[unique_node_indices[i], 'mu']
      }
    }
    # More here to deal with more complicated trees - i.e. multiple trees
  } else {
    # Do a recursive call to the function
    partial_trees = trees
    partial_trees[[1]] = NULL # Blank out that element of the list
    predictions = get_predictions(trees[[1]], X, single_tree = TRUE)  +
      get_predictions(partial_trees, X,
                      single_tree = length(partial_trees) == 1)
    #single_tree = !is.null(names(partial_trees)))
    # The above only sets single_tree to if the names of the object is not null (i.e. is a list of lists)
  }

  return(predictions)
}

# get_children ------------------------------------------------------------

get_children = function(tree_mat, parent) {
  # Create a holder for the children
  all_children = NULL
  if(as.numeric(tree_mat[parent,'terminal']) == 1) {
    # If the node is terminal return the list so far
    return(c(all_children, parent))
  } else {
    # If not get the current children
    curr_child_left = as.numeric(tree_mat[parent, 'child_left'])
    curr_child_right = as.numeric(tree_mat[parent, 'child_right'])
    # Return the children and also the children of the children recursively
    return(c(all_children,
             get_children(tree_mat,curr_child_left),
             get_children(tree_mat,curr_child_right)))
  }
}

# Sample function ----------------------------------------------------------

resample <- function(x, ...) x[sample.int(length(x), size=1), ...]

update_s = function(var_count, p, alpha_s){
  s_ = rdirichlet(1, alpha_s/p + var_count)
  return(s_)
}

# get_number_distinct_cov <- function(tree){
#
#   # Select the rows that correspond to internal nodes
#   which_terminal = which(tree$tree_matrix[,'terminal'] == 0)
#   # Get the covariates that are used to define the splitting rules
#   num_distinct_cov = length(unique(tree$tree_matrix[which_terminal,'split_variable']))
#
#   return(num_distinct_cov)
# }

sample_move = function(curr_tree, i, nburn){

  if (nrow(curr_tree$tree_matrix) == 1 || i < max(floor(0.1*nburn), 10)) {
    type = 'grow'
  } else {
    type = sample(c('grow', 'prune', 'change'), 1)
  }
  return(type)
}


R/predict_tree.R
#' @export
predict_mybart = function(object, newdata,
                         type = c('all', 'median', 'mean')) {



  for(i in 1:ncol(newdata)) {
    newdata[,i] <- object$ecdfs[[i]](newdata[,i])
  }

  # Create holder for predicted values
  n_newX = dim(newdata)[1]
  n_its = object$npost
  y_hat_mat = matrix(NA, nrow = n_its,
                     ncol = nrow(newdata))

  # Now loop through iterations and get predictions
  for (i in 1:n_its) {
    # Get current set of trees
    curr_trees = object$trees[[i]]

    # Use get_predictions function to get predictions
    y_hat_mat[i,] = get_predictions(curr_trees,
                                    newdata,
                                    single_tree = length(curr_trees) == 1)
  }

  # Sort out what to return
  out = switch(type,
               all = object$y_mean + object$y_sd * y_hat_mat,
               mean = object$y_mean + object$y_sd * apply(y_hat_mat,2,'mean'),
               median = object$y_mean + object$y_sd * apply(y_hat_mat,2,'median'))

  return(out)

} # end of predict function
#' @export
marginal_predict_mybart = function(object, var_marg, newdata,
                               type = c('all', 'median', 'mean')) {

  for(i in 1:ncol(newdata)) {
    newdata[,i] <- object$ecdfs[[i]](newdata[,i])
  }

  # Create holder for predicted values
  n_newX = dim(newdata)[1]
  n_its = object$npost
  ntrees = object$ntrees
  y_hat_mat = matrix(0, nrow = n_its,
                     ncol = nrow(newdata))

  # Get which covariates are used by each tree in each MCMC iteration
  vars_trees = matrix(NA, nrow=n_its, ncol=ntrees)
  for (i in 1:n_its){
    for (j in 1:ntrees){
      aux = object$trees[[i]][[j]]$tree_matrix[,'split_variable']
      if(length(aux) > 1){
        vars_trees[i,j] = paste(unique(sort(aux[!is.na(aux)])), collapse = ',')
      }
    }
  }

  # Now loop through iterations and get predictions
  for (i in 1:n_its) {
    marginal_trees = which(vars_trees[i,] == var_marg)
    # Sometimes the trees do not contain the variable we're interested in
    if (length(marginal_trees) > 0){
      # Get current set of trees
      curr_trees = object$trees[[i]][marginal_trees]

      # Use get_predictions function to get predictions
      y_hat_mat[i,] = get_predictions(curr_trees,
                                      newdata,
                                      single_tree = length(curr_trees) == 1)

    }
  }

  # Sort out what to return
  out = list(vars_trees = vars_trees,
             prediction = switch(type,
                                 all = object$y_mean + object$y_sd * y_hat_mat,
                                 mean = object$y_mean + object$y_sd * apply(y_hat_mat,2,'mean'),
                                 median = object$y_mean + object$y_sd * apply(y_hat_mat,2,'median')))

  return(out)

} # end of predict function


########################################################################################################
# Predictions for classification
########################################################################################################

#' @export
predict_mybart_class = function(object,
                                   newdata,
                                   type = c('all', 'median', 'mean')) {

  # Create holder for predicted values
  n_newX = dim(newdata)[1]
  n_its = object$npost
  y_hat_mat = matrix(NA, nrow = n_its,
                     ncol = nrow(newdata))

  # Now loop through iterations and get predictions
  for (i in 1:n_its) {
    # Get current set of trees
    curr_trees = object$trees[[i]]

    # Use get_predictions function to get predictions
    y_hat_mat[i,] = get_predictions(curr_trees,
                                    newdata,
                                    single_tree = length(curr_trees) == 1)
  }

  # Sort out what to return
  out = switch(type,
               all = y_hat_mat,
               mean = apply(pnorm(y_hat_mat),2,'mean'),
               median = apply(pnorm(y_hat_mat),2,'median'))

  return(out)

} # end of predict function


#' @export
marginal_predict_mybart_class = function(object, var_marg, newdata,
                                   type = c('all', 'median', 'mean')) {

  # Create holder for predicted values
  n_newX = dim(newdata)[1]
  n_its = object$npost
  ntrees = object$ntrees
  y_hat_mat = matrix(0, nrow = n_its,
                     ncol = nrow(newdata))

  # Get which covariates are used by each tree in each MCMC iteration
  vars_trees = matrix(NA, nrow=n_its, ncol=ntrees)
  for (i in 1:n_its){
    for (j in 1:ntrees){
      aux = object$trees[[i]][[j]]$tree_matrix[,'split_variable']
      if(length(aux) > 1){
        vars_trees[i,j] = paste(unique(sort(aux[!is.na(aux)])), collapse = ',')
      }
    }
  }

  # Now loop through iterations and get predictions
  for (i in 1:n_its) {
    marginal_trees = which(vars_trees[i,] == var_marg)
    # Sometimes the trees do not contain the variable we're interested in
    if (length(marginal_trees) > 0){
      # Get current set of trees
      curr_trees = object$trees[[i]][marginal_trees]

      # Use get_predictions function to get predictions
      y_hat_mat[i,] = get_predictions(curr_trees,
                                      newdata,
                                      single_tree = length(curr_trees) == 1)

    }
  }

  # Sort out what to return
  out = list(vars_trees = vars_trees,
             prediction = switch(type,
                                 all    = pnorm(y_hat_mat),
                                 mean   = apply(pnorm(y_hat_mat),2,'mean'),
                                 median = apply(pnorm(y_hat_mat),2,'median')))

  return(out)

} # end of predict function


R/parameters_quantities_tree.R
# -------------------------------------------------------------------------#
# Description: this script contains 2 functions that are used to generate  #
#              the predictions, update variance and compute the tree prior #
#              and the marginalised likelihood                             #
# -------------------------------------------------------------------------#

# 1. simulate_mu: generate the predicted values (mu's)
# 2. update_sigma2: updates the parameters sigma2
# 3. update_z: updates the latent variables z. This is required for MOTR-BART for classification.
# 4. get_tree_prior: returns the tree log prior score
# 5. tree_full_conditional: computes the marginalised likelihood for all nodes for a given tree
# 6. get_number_distinct_cov: counts the number of distinct covariates that are used in a tree to create the splitting rules
# Compute the full conditionals -------------------------------------------------

tree_full_conditional = function(tree, R, sigma2, sigma2_mu) {

  # Function to compute log full conditional distirbution for an individual tree
  # R is a vector of partial residuals

  # Need to calculate log complete conditional, involves a sum over terminal nodes

  # First find which rows are terminal nodes
  which_terminal = which(tree$tree_matrix[,'terminal'] == 1)

  # Get node sizes for each terminal node
  nj = tree$tree_matrix[which_terminal,'node_size']

  # Get sum of residuals and sum of residuals squared within each terminal node
  sumRsq_j = aggregate(R, by = list(tree$node_indices), function(x) sum(x^2))[,2]
  S_j = aggregate(R, by = list(tree$node_indices), sum)[,2]


  # print(" tree$node_indices = ")
  # print(tree$node_indices)
  #
  # print("unique(tree$node_indices) = ")
  # print(unique(tree$node_indices))
  #
  # print("nj = ")
  # print(nj)
  #
  # print("S_j = ")
  # print(S_j)


  if(length(nj) != length(S_j)){
    print("tree = ")
    print(tree)
    stop("length(nj) != length(S_j)")
  }

  # print("sigma2 = ")
  # print(sigma2)
  #
  # print("sumRsq_j = ")
  # print(sumRsq_j)
  #
  # print("R = ")
  # print(R)


  # Now calculate the log posterior
  log_post = 0.5 * ( sum(log( sigma2 / (nj*sigma2_mu + sigma2))) +
              sum( (sigma2_mu* S_j^2) / (sigma2 * (nj*sigma2_mu + sigma2))))

  return(log_post)
}


# Simulate_par -------------------------------------------------------------

simulate_mu = function(tree, R, sigma2, sigma2_mu) {

  # Simulate mu values for a given tree

  # First find which rows are terminal nodes
  which_terminal = which(tree$tree_matrix[,'terminal'] == 1)

  # Get node sizes for each terminal node
  nj = tree$tree_matrix[which_terminal,'node_size']

  # Get sum of residuals in each terminal node
  sumR = aggregate(R, by = list(tree$node_indices), sum)[,2]


  # print(" tree$node_indices = ")
  # print(tree$node_indices)
  #
  # print("nj = ")
  # print(nj)
  #
  # print("sumR = ")
  # print(sumR)

  if(length(nj) != length(sumR)){
    print("tree = ")
    print(tree)
    stop("length(nj) != length(sumR)")
  }

  # print("sigma2 = ")
  # print(sigma2)
  #
  # print("sigma2_mu = ")
  # print(sigma2_mu)

  # Now calculate mu values
  mu = rnorm(length(nj) ,
             mean = (sumR / sigma2) / (nj/sigma2 + 1/sigma2_mu),
             sd = sqrt(1/(nj/sigma2 + 1/sigma2_mu)))

  # Wipe all the old mus out for other nodes
  tree$tree_matrix[,'mu'] = NA

  # Put in just the ones that are useful
  tree$tree_matrix[which_terminal,'mu'] = mu

  return(tree)
}

# Update sigma2 -------------------------------------------------------------

update_sigma2 <- function(S, n, nu, lambda){
  u = 1/rgamma(1, shape = (n + nu)/2, rate = (S + nu*lambda)/2)
  return(u)
}

# Update the latent variable z ---------------

update_z = function(y, prediction){

  ny0 = sum(y==0)
  ny1 = sum(y==1)
  z = rep(NA, length(y))

  z[y==0] = rtruncnorm(ny0, a = -Inf, b=0,   mean = prediction[y==0], 1)
  z[y==1] = rtruncnorm(ny1, a = 0   , b=Inf, mean = prediction[y==1], 1)

  return(z)
}

# Get tree priors ---------------------------------------------------------

  get_tree_prior = function(tree, alpha, beta) {

  # Need to work out the depth of the tree
  # First find the level of each node, then the depth is the maximum of the level
  level = rep(NA, nrow(tree$tree_matrix))
  level[1] = 0 # First row always level 0

  # Escpae quickly if tree is just a stump
  if(nrow(tree$tree_matrix) == 1) {
    return(log(1 - alpha)) # Tree depth is 0
  }

  for(i in 2:nrow(tree$tree_matrix)) {
    # Find the current parent
    curr_parent = as.numeric(tree$tree_matrix[i,'parent'])
    # This child must have a level one greater than it's current parent
    level[i] = level[curr_parent] + 1
  }

  # Only compute for the internal nodes
  internal_nodes = which(as.numeric(tree$tree_matrix[,'terminal']) == 0)
  log_prior = 0
  for(i in 1:length(internal_nodes)) {
    log_prior = log_prior + log(alpha) - beta * log(1 + level[internal_nodes[i]])
  }
  # Now add on terminal nodes
  terminal_nodes = which(as.numeric(tree$tree_matrix[,'terminal']) == 1)
  for(i in 1:length(terminal_nodes)) {
    log_prior = log_prior + log(1 - alpha * ((1 + level[terminal_nodes[i]])^(-beta)))
  }

  return(log_prior)

  }

# get_num_cov_prior <- function(tree, lambda_cov, nu_cov, penalise_num_cov){
#
#   # Select the rows that correspond to internal nodes
#   which_terminal = which(tree$tree_matrix[,'terminal'] == 0)
#   # Get the covariates that are used to define the splitting rules
#   num_distinct_cov = length(unique(tree$tree_matrix[which_terminal,'split_variable']))
#
#   if (penalise_num_cov == TRUE){
#     if (num_distinct_cov > 0){
#       # A zero-truncated double Poisson
#       log_prior_num_cov = ddoublepois(num_distinct_cov, lambda_cov, nu_cov, log = TRUE) -
#         log(1-ddoublepois(0,lambda_cov, nu_cov))
#     } else {
#       log_prior_num_cov = 0
#     }
#   } else {
#     log_prior_num_cov = 0 # no penalisation
#   }
#
#   return(log_prior_num_cov)
#
# }


  #The get w function determine the number of second generation internal nodes in a given tree, input: tree
  get_w = function(tree){
    if(is.null(tree) | nrow(tree$tree_matrix) == 1) { # In case the tree is empty or a stump
      B = 0
    } else {
      indeces = which(tree$tree_matrix[,'terminal'] == '1') #determine which indexes have a terminal node label
      # determine the parent for each terminal node and sum the number of duplicated parents
      B = as.numeric(sum(duplicated(tree$tree_matrix[indeces,'parent'])))}
    return(B)
  }

  # The get number of terminals function calculates the number of terminal nodes, input: tree
  get_nterminal = function(tree){
    if(is.null(tree)) { # In case the tree is empty
      L = 0
    } else {
      indeces = which(tree$tree_matrix[,'terminal'] == '1')  # determine which indexes have a terminal node label
      L = as.numeric(length(indeces))} # take the length of these indexes, to determine the number of terminal nodes/leaves
    return(L)
  }

  # The ratio grow function calculates ratio of transition probabilities of the grow step
  # This function is in accordance with the soft BART paper, input: current tree, proposed/new tree
  ratio_grow = function(curr_tree,new_tree){
    grow_ratio = get_nterminal(curr_tree)/(get_w(new_tree)+1)
    return(grow_ratio)
  }

  # The ratio prune function calculates ratio of transition probabilities of the prune step
  # This function is in accordance with the soft BART paper, input: current tree, proposed/new tree
  ratio_prune = function(curr_tree,new_tree){
    prune_ratio = get_w(new_tree)/(get_nterminal(curr_tree)-1)
    return(prune_ratio)
  }

  # The alpha MH function calculates the alpha of the Metropolis-Hastings step based on the type of transformation
  # input: transformation type, current tree, proposed/new tree and the respective likelihoods
  alpha_mh = function(l_new,l_old, curr_trees,new_trees, type){
    if(type == 'grow'){
      a = exp(l_new - l_old)*ratio_grow(curr_trees, new_trees)

    } else if(type == 'prune'){

      a = exp(l_new - l_old)*ratio_prune(curr_trees, new_trees)
    } else{
      a = exp(l_new - l_old)
    }
    return(a)
  }


R/myBART.R

#' @title Bayesian Additive Regression Trees with Oblique Splits
#'
#' @description Bayesian Additive Regression Trees with Oblique Splits
#' @importFrom mvtnorm 'rmvnorm'
#' @importFrom stats 'rgamma' 'runif' 'dnorm' 'sd' 'rnorm' 'pnorm' 'aggregate' 'contrasts' 'model.matrix'
#' @importFrom MCMCpack 'rdirichlet'
#' @importFrom truncnorm 'rtruncnorm'
#' @param x Training covariate matrix. Rows correspond to observations, columns correspond to covariates used in splitting rules.
#' @param y Training outcome vector.
#' @return The following objects are returned:
#' \item{trees}{List (over MCMC iterations) of lists (over trees in a sum-of-trees model) of tree matrices.}
#'
#' @useDynLib ObliqueBART, .registration = TRUE
#' @export


# x
# y
# sparse = FALSE
# ntrees = 10
# node_min_size = 5
# alpha = 0.95
# beta = 2
# nu = 3
# lambda = 0.1
# mu_mu = 0
# sigma2 = 1
# sigma2_mu = 1
# nburn = 1000
# npost = 1000
# nthin = 1
# lambda_cov = 0.4
# nu_cov = 2
# penalise_num_cov = FALSE

ObliqueBART <- function(   x,
                   y,
                   sparse = TRUE,
                   ntrees = 10,
                   node_min_size = 5,
                   alpha = 0.95,
                   beta = 2,
                   nu = 3,
                   lambda = 0.1,
                   mu_mu = 0,
                   sigma2 = 1,
                   sigma2_mu = 1,
                   nburn = 1000,
                   npost = 1000,
                   nthin = 1,
                   penalise_num_cov = TRUE,
                   lambda_cov = 0.4,
                   nu_cov = 2) {

  # Extract control parameters
  node_min_size = node_min_size

  # Extract MCMC details
  TotIter = nburn + npost*nthin # Total of iterations

  # Storage containers
  store_size = npost
  tree_store = vector('list', store_size)
  sigma2_store = rep(NA, store_size)
  y_hat_store = matrix(NA, ncol = length(y), nrow = store_size)
  var_count = rep(0, ncol(x))
  var_count_store = matrix(0, ncol = ncol(x), nrow = store_size)
  s_prob_store = matrix(0, ncol = ncol(x), nrow = store_size)
  tree_fits_store = matrix(0, ncol = ntrees, nrow = length(y))

  # Scale the response target variable
  y_mean = mean(y)
  y_sd = sd(y)
  y_scale = (y - y_mean)/y_sd
  n = length(y_scale)
  p = ncol(x)
  s = rep(1/p, p)

  # Create a list of trees for the initial stump
  curr_trees = create_stump(num_trees = ntrees,
                            y = y_scale,
                            X = x)
  # Initialise the new trees as current one
  new_trees = curr_trees

  # Initialise the predicted values to zero
  y_hat = get_predictions(curr_trees, x, single_tree = ntrees == 1)

  # Set up a progress bar
  pb = utils::txtProgressBar(min = 1, max = TotIter,
                             style = 3, width = 60,
                             title = 'Running rBART...')

  # Start the MCMC iterations loop
  for (i in 1:TotIter) {

    utils::setTxtProgressBar(pb, i)

    # If at the right place, store everything
    if((i > nburn) & ((i - nburn) %% nthin) == 0) {
      curr = (i - nburn)/nthin
      tree_store[[curr]] = curr_trees
      sigma2_store[curr] = sigma2
      y_hat_store[curr,] = y_hat
      var_count_store[curr,] = var_count
      s_prob_store[curr,] = s
    }

      # Start looping through trees
      for (j in 1:ntrees) {

        current_partial_residuals = y_scale - y_hat + tree_fits_store[,j]

        # Propose a new tree via grow/change/prune/swap
        type = sample_move(curr_trees[[j]], i, nburn)

        # Generate a new tree based on the current
        new_trees[[j]] = update_tree(y = y_scale,
                                     X = x,
                                     type = type,
                                     curr_tree = curr_trees[[j]],
                                     node_min_size = node_min_size,
                                     s = s)

        # CURRENT TREE: compute the log of the marginalised likelihood + log of the tree prior
        l_old = tree_full_conditional(curr_trees[[j]],
                                      current_partial_residuals,
                                      sigma2,
                                      sigma2_mu) +
          get_tree_prior(curr_trees[[j]], alpha, beta)

        # NEW TREE: compute the log of the marginalised likelihood + log of the tree prior
        l_new = tree_full_conditional(new_trees[[j]],
                                      current_partial_residuals,
                                      sigma2,
                                      sigma2_mu) +
          get_tree_prior(new_trees[[j]], alpha, beta)

        # Exponentiate the results above
        a = exp(l_new - l_old)

        if(a > runif(1)) {
          curr_trees[[j]] = new_trees[[j]]
        }

        # Update mu whether tree accepted or not
        curr_trees[[j]] = simulate_mu(curr_trees[[j]],
                                      current_partial_residuals,
                                      sigma2,
                                      sigma2_mu)
      # Updating BART predictions
      current_fit = get_predictions(curr_trees[j], x, single_tree = TRUE)
      y_hat = y_hat - tree_fits_store[,j] # subtract the old fit
      y_hat = y_hat + current_fit # add the new fit
      tree_fits_store[,j] = current_fit # update the new fit

      } # End loop through trees

    # y_hat = get_predictions(curr_trees, x, single_tree = ntrees == 1)
    sum_of_squares = sum((y_scale - y_hat)^2)

    # Update sigma2 (variance of the residuals)
    sigma2 = update_sigma2(sum_of_squares, n = length(y_scale), nu, lambda)

    # Update s = (s_1, ..., s_p), where s_p is the probability that predictor p is used to create new terminal nodes
    if (sparse == 'TRUE' & i > floor(TotIter*0.1)){
      s = update_s(var_count, p, 1)
    }
  } # End iterations loop

  cat('\n') # Make sure progress bar ends on a new line

  return(list(trees = tree_store,
              sigma2 = sigma2_store*y_sd^2,
              y_hat = y_hat_store*y_sd + y_mean,
              npost = npost,
              nburn = nburn,
              nthin = nthin,
              ntrees = ntrees,
              y_mean = y_mean,
              y_sd = y_sd,
              var_count_store = var_count_store,
              s = s_prob_store
              ))

} # End main function

